man/fi_mr.3.md

```diff
@@ -287,27 +264,64 @@ The following apply to memory registration.
   MR with an endpoint, the application must use fi_mr_bind().  To
   enable the memory region, the application must call fi_mr_enable().
 
-*FI_MR_HMEM*
-: This mode bit is associated with the FI_HMEM capability.
-  If FI_MR_HMEM is set, the application must register buffers that
-  were allocated using a device call and provide a valid desc
-  parameter into applicable data transfer operations even if they are
-  only used for local operations (e.g. send and receive data buffers).
-  Device memory must be registered using the fi_mr_regattr call, with
-  the iface and device fields filled out.
-
-  If FI_MR_HMEM is set, but FI_MR_LOCAL is unset, only device buffers
-  must be registered when used locally.  In this case, the desc parameter
-  passed into data transfer operations must either be valid or NULL.
-  Similarly, if FI_MR_LOCAL is set, but FI_MR_HMEM is not, the desc
-  parameter must either be valid or NULL.
-
-*FI_MR_COLLECTIVE*
-: This bit is associated with the FI_COLLECTIVE capability.  When set,
-  the provider requires that memory regions used in collection operations
-  must explicitly be registered for use with collective calls.  This
-  requires registering regions passed to collective calls using the
-  FI_COLLECTIVE flag.
+*Explicit Memory Registration Requirements*
+: Depending on memory and/or operation type, a provider may have different
+  requirements around explicit memory registration. Explicit memory
+  registration means the application must register data buffers being used for
+  the local operation (e.g. send and receive). Three different MR modes bits
+  are defined for this: FI_MR_LOCAL, FI_MR_HMEM, and FI_MR_COLLECTIVE.
```

Comment:
FI_MR_HMEM and FI_MR_COLLECTIVE are independent from FI_MR_LOCAL.

FI_MR_COLLECTIVE is **not** restricted to local operations only.  It indicates that the buffers must be registered specifically for use with any collective operation.  The registration call must include the FI_COLLECTIVE flag.

FI_MR_HMEM is not quite a local memory registration requirement for device memory.  Even if the buffer is only access by an RDMA write, for example, FI_MR_HMEM requires that the buffer be registered using the hmem registration semantics.  This is true even in cases of unified addressing, where device memory shares the same VA space as system memory.

Please don't group these 3 bits together under the same definition.

Reply:
Will fix.